documentation for workspaces (#1086)
conan-io/docs: developing_packages/workspaces.rst

a different layout per package, as:

.. code-block:: yaml

    editables:
        say/0.1@user/testing:
            path: say
            layout: custom_say_layout

Layout files are explained in :ref:`editable_layout` and in the :ref:`editable_packages` sections.

The ``workspace_generator`` defines the file that will be generated for the top project. The only supported value so far
is ``cmake`` and it will generate a *conanworkspace.cmake* file that looks like:

.. code-block:: cmake

    set(PACKAGE_say_SRC "<path>/examples/workspace/cmake/say/src")
    set(PACKAGE_say_BUILD "<path>/examples/workspace/cmake/say/build/Debug")
    set(PACKAGE_hello_SRC "<path>/examples/workspace/cmake/hello/src")
    set(PACKAGE_hello_BUILD "<path>/examples/workspace/cmake/hello/build/Debug")
    set(PACKAGE_chat_SRC "<path>/examples/workspace/cmake/chat/src")
    set(PACKAGE_chat_BUILD "<path>/examples/workspace/cmake/chat/build/Debug")

    macro(conan_workspace_subdirectories)
        add_subdirectory(${PACKAGE_say_SRC} ${PACKAGE_say_BUILD})
        add_subdirectory(${PACKAGE_hello_SRC} ${PACKAGE_hello_BUILD})
        add_subdirectory(${PACKAGE_chat_SRC} ${PACKAGE_chat_BUILD})
    endmacro()

> This file can be included in your root *CMakeLists.txt* (this one is not generated, you provide it, so you can customize it

Reviewer: I would rephrase this two lines to something like:

> The consumption of the packages is done including this file in an user-defined *CMakeLists.txt* file so you can customize it as much as you want. Here you can see a simple one included in this project:

    $ conan search
    There are no packages

.. note::

    The package *conanfile.py* recipes do not contain anything special, they are standard recipes. But the packages *CMakeLists.txt*
    have defined the following:

    .. code-block:: cmake

        conan_basic_setup(NO_OUTPUT_DIRS)

>     This is because the default ``conan_basic_setup()`` does define output directories for artifacts as *bin*, *lib*, etc, which is

Reviewer 2: Why not leaving the standard `conan_basic_setup()` call and then make the `layout` file match to it? Explanation, if needed, will be closer to the layout file description, not to the CMake one.

Author: Good point, I'll try to improve it later, but I am going to submit my review, so it can be merged soon.

.. code-block:: yaml

    editables:
    say/0.1@user/testing:
        path: say
    hello/0.1@user/testing:
        path: hello
    chat/0.1@user/testing:
        path: chat
    layout: layout_vs
    generators: cmake_multi
    workspace_generator: cmake
    root: chat/0.1@user/testing

Note the ``generators: cmake_multi`` line, that will define the generators to be used by our workspace packages. Also, our *CMakeLists.txt*
should take into account that now we won't have a *conanbuildinfo.cmake* file, but a *conanbuildinfo_multi.cmake* file. See for example
the *hello/src/CMakeLists.txt* file:

.. code-block:: cmake

    project(Hello)

    if(EXISTS ${CMAKE_CURRENT_BINARY_DIR}/conanbuildinfo_multi.cmake)
        include(${CMAKE_CURRENT_BINARY_DIR}/conanbuildinfo_multi.cmake)
    else()
        include(${CMAKE_CURRENT_BINARY_DIR}/conanbuildinfo.cmake)
    endif()

    conan_basic_setup(NO_OUTPUT_DIRS)

    add_library(hello hello.cpp)
    conan_target_link_libraries(hello)

The last ``conan_target_link_libraries(hello)`` is a helper that does the right linking with Debug/Release libraries (also works when using cmake
targets).

> Lets use this. Note it is very important to install both Debug and Release configurations straight ahead, if we want to later switch between them

Reviewer: ```suggestion
Make sure you install both Debug and Release configurations straight ahead, so we can switch build type later
```

in the IDE:

.. code-block:: bash

    $ mkdir build && cd build
    $ conan workspace install ../conanws_vs.yml
    $ conan workspace install ../conanws_vs.yml -s build_type=Debug
    $ cmake .. -G "Visual Studio 15 Win64"

With those commands you will get a Visual Studio solution, that you can open, select the *app* executable as StartUp project, and start building,
executing, debugging, switching from Debug and Release configurations freely from the IDE, without needing to issue further Conan commands.

You can check in the project folders, how the following files have been generated:

.. code-block:: text

    hello
      |- build
            | - conanbuildinfo_multi.cmake
            | - conanbuildinfo_release.cmake
            | - conanbuildinfo_debug.cmake

Note that they are not located in *build/Release* and *build/Debug* subfolders, that is because of the multi-config environment. To account for that
the *layout_vs* define the ``[build_folder]`` not as ``build/{settings.build_type}`` but just as:

.. code-block:: ini

    [build_folder]
    build

Notes
-----

> Not that this way of developing packages shouldn't be used to create the final packages (you could try to use ``conan export-pkg``), but instead,

Reviewer: ```suggestion
Note that this way of developing packages shouldn't be used to create the final packages (you could try to use :command:`conan export-pkg`). Instead,
```

> a full package creation with ``conan create`` (best in CI) is recommended. 

Reviewer: ```suggestion
a full package creation with :command:`conan create` is recommended. 
```
